cpp/src/copying/contiguous_split.cu

```diff
@@ -1158,116 +1286,700 @@ std::vector<packed_table> contiguous_split(cudf::table_view const& input,
+      std::move(std::vector<std::size_t>{total_size}),
+      total_size);
+  }
+}
+
+void copy_data(int num_batches_to_copy,
+               int starting_batch,
+               uint8_t const** d_src_bufs,
+               uint8_t** d_dst_bufs,
+               rmm::device_uvector<dst_buf_info>& d_dst_buf_info,
+               uint8_t* user_buffer,
+               rmm::cuda_stream_view stream)
+{
+  constexpr size_type block_size = 256;
+  if (user_buffer != nullptr) {
+    copy_partitions<block_size><<<num_batches_to_copy, block_size, 0, stream.value()>>>(
+      d_src_bufs, user_buffer, d_dst_buf_info.data() + starting_batch);
+  } else {
+    copy_partitions<block_size><<<num_batches_to_copy, block_size, 0, stream.value()>>>(
+      d_src_bufs, d_dst_bufs, d_dst_buf_info.data() + starting_batch);
+  }
+}
+
+/**
+ * @brief Function that checks an input table_view and splits for specific edge cases.
+ *
+ * It will return true if the input is "empty" (no rows or columns), which means
+ * special handling has to happen in the calling code.
+ *
+ * @param input table_view of source table to be split
+ * @param splits the splits specified by the user, or an empty vector if no splits
+ * @returns true if the input is empty, false otherwise
+ */
+bool check_inputs(cudf::table_view const& input, std::vector<size_type> const& splits)
+{
+  if (input.num_columns() == 0) { return true; }
+  if (splits.size() > 0) {
+    CUDF_EXPECTS(splits.back() <= input.column(0).size(),
+                 "splits can't exceed size of input columns");
+  }
+  {
+    size_type begin = 0;
+    for (std::size_t i = 0; i < splits.size(); i++) {
+      size_type end = splits[i];
+      CUDF_EXPECTS(begin >= 0, "Starting index cannot be negative.");
+      CUDF_EXPECTS(end >= begin, "End index cannot be smaller than the starting index.");
+      CUDF_EXPECTS(end <= input.column(0).size(), "Slice range out of bounds.");
+      begin = end;
+    }
+  }
+  return input.column(0).size() == 0;
+}
+
+/**
+ * @brief A helper struct containing the state of contiguous_split, whether the caller
+ * is using the single-pass contiguous_split or chunked_pack.
+ *
+ * It exposes an iterator-like pattern where contiguous_split_state::has_next()
+ * return true when there is work to be done, and false otherwise.
+ *
+ * contiguous_split_state::contiguous_split() performs a single-pass contiguous_split
+ * and is only valid iff contiguous_split_state is instantiated with 0 for the user_buffer_size.
+ *
+ * contiguous_split_state::contiguous_split_chunk(device_span) is only valid when
+ * user_buffer_size > 0. It should be called as long as has_next() returns true. The
+ * device_span passed to contiguous_split_chunk must be allocated in stream `stream` by
+ * the user.
+ *
+ * None of the methods are thread safe.
+ */
+struct contiguous_split_state {
+  static const std::size_t desired_batch_size = 1 * 1024 * 1024;
```

Comment:
I feel like this value needs to be in a more prominent place.  It's super crucial to the whole operation.